### PR TITLE
Added Rancher Test case: NamespaceMoveFromProjectToProjectToNoProjectWhileRestore

### PR DESCRIPTION
**Pull request by snigdha-px**
**What this PR does / why we need it**:
Added New Rancher testcase
Added retry in cluster delete in AfterSuite as RKE cluster takes time to delete. If cluster is not deleted, corresponding cloud cred deletion fails
**Which issue(s) this PR fixes** (optional)
Closes #https://portworx.atlassian.net/browse/PA-1497

**Special notes for your reviewer**:
Log : https://jenkins.pwx.dev.purestorage.com/job/portworx-backup/job/system-tests/job/byoc/job/px-backup-on-demand-system-test-byoc/2175/consoleFull

### CODE REVIEW

File: drivers/scheduler/rke/rke.go [hunk 1]
```diff
@@ -270,6 +270,53 @@ func (r *Rancher) SaveSchedulerLogsToFile(n node.Node, location string) error {
 	return err
 }
 
+// MoveNamespaceToNoProject moves namespace to no project
+func (r *Rancher) MoveNamespaceToNoProject(nsList []string) error {
```
Comment (mkoppal-px) [inline drivers/scheduler/rke/rke.go:274]: Can we call this function `RemoveNamespaceFromProject`?

File: drivers/scheduler/rke/rke.go [hunk 2]
```diff
@@ -270,6 +270,53 @@ func (r *Rancher) SaveSchedulerLogsToFile(n node.Node, location string) error {
...
+		_, err = core.Instance().UpdateNamespace(ns)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// MoveNamespaceFromOneProjectToAnotherProject moves namespace from one project to other project
+func (r *Rancher) MoveNamespaceFromOneProjectToAnotherProject(projectName string, nsList []string) error {
```
Comment (mkoppal-px) [inline drivers/scheduler/rke/rke.go:291]: Can we call this `ChangeProjectForNamespace`?

File: tests/backup/backup_basic_test.go [hunk 3]
```diff
@@ -254,9 +254,23 @@ var _ = AfterSuite(func() {
 				}
 				err = DeleteCluster(clusterName, orgID, ctx, true)
 				Inst().Dash.VerifySafely(err, nil, fmt.Sprintf("Deleting cluster %s", clusterName))
-				if clusterCredName != "" {
-					err = DeleteCloudCredential(clusterCredName, orgID, clusterCredUID)
-					Inst().Dash.VerifySafely(err, nil, fmt.Sprintf("Verifying deletion of cluster cloud cred [%s]", clusterCredName))
+				if err == nil {
```
Comment (mkoppal-px) [inline tests/backup/backup_basic_test.go:257]: Why do you have to check for err?
We can do the same thing which you are doing in the other cleanup right?

```
				clusterDeleteStatus := func() (interface{}, bool, error) {
					status, err := IsClusterPresent(clusterName, ctx, orgID)
					if err != nil {
						return "", true, fmt.Errorf("cluster %s still present with error %v", clusterName, err)
					}
					if status {
						return "", true, fmt.Errorf("cluster %s is not deleted yet", clusterName)
					}
					return "", false, nil
				}
				_, err = task.DoRetryWithTimeout(clusterDeleteStatus, clusterDeleteTimeout, clusterDeleteRetryTime)
				Inst().Dash.VerifySafely(err, nil, fmt.Sprintf("Deleting cluster %s", clusterName))
```

File: tests/backup/backup_helper.go [hunk 4]
```diff
@@ -3910,6 +3911,50 @@ func CreateRestoreWithProjectMapping(restoreName string, backupName string, name
 	return nil
 }
 
+// CreateRestoreWithProjectMappingWithoutCheckWithReplacePolicy creates restore with project mapping
+func CreateRestoreWithProjectMappingWithoutCheckWithReplacePolicy(restoreName string, backupName string, namespaceMapping map[string]string, clusterName string,
```
Comment (mkoppal-px) [inline tests/backup/backup_helper.go:3915]: Can we simply call this `CreateRestoreOnRancherWithoutCheck` ?

File: drivers/scheduler/rke/rke.go (hunk 1 above)
Comment (snigdha-px) [reply drivers/scheduler/rke/rke.go:274]: done

File: drivers/scheduler/rke/rke.go (hunk 2 above)
Comment (snigdha-px) [reply drivers/scheduler/rke/rke.go:291]: done

File: tests/backup/backup_helper.go (hunk 4 above)
Comment (snigdha-px) [reply tests/backup/backup_helper.go:3915]: done

File: tests/backup/backup_basic_test.go (hunk 3 above)
Comment (snigdha-px) [reply tests/backup/backup_basic_test.go]: Yes correct. Don't know how I missed it.
Corrected now

File: tests/backup/backup_rancher_test.go [hunk 5]
```diff
@@ -265,3 +266,211 @@ var _ = Describe("{SingleNamespaceBackupRestoreToNamespaceInSameAndDifferentProj
...
+			err = Inst().S.(*rke.Rancher).ValidateProjectOfNamespaces(sourceProject, appNamespaces)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying if namespace %s is added to project %s", appNamespaces, sourceProject))
+			log.InfoD("Taking Backup of applications %s", appNamespaces)
+			backupName = fmt.Sprintf("%s-%v", BackupNamePrefix, RandomString(10))
+			appContextsToBackup := FilterAppContextsByNamespace(scheduledAppContexts, appNamespaces)
+			err = CreateBackupWithValidation(ctx, backupName, SourceClusterName, customBackupLocationName, backupLocationUID, appContextsToBackup, nil, orgID, sourceClusterUid, "", "", "", "")
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Creation and Validation of backup [%s]", backupName))
+		})
+
+		Step("Creating 2 rancher project in destination cluster", func() {
```
Comment (mkoppal-px) [inline tests/backup/backup_rancher_test.go:371]: nit: `projects`

Comment (snigdha-px) [reply tests/backup/backup_rancher_test.go:371]: done

File: tests/backup/backup_rancher_test.go [hunk 6]
```diff
@@ -265,3 +266,211 @@ var _ = Describe("{SingleNamespaceBackupRestoreToNamespaceInSameAndDifferentProj
...
+			restoreName = fmt.Sprintf("%s-replace-ns-move-to-no-project-%v", restoreNamePrefix, RandomString(5))
+			restoreList = append(restoreList, restoreName)
+			err = CreateRestoreOnRancherWithoutCheck(restoreName, backupName, namespaceMapping, destinationClusterName, orgID, ctx, nil, projectUIDMapping, projectNameMapping, 2)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Creating restore- %s from backup %s", restoreName, backupName))
+			log.Infof("Move destination ns to no project while restore is going on")
+			err = Inst().S.(*rke.Rancher).RemoveNamespaceFromProject(restoreNamespaceList)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Moving namespaces %v from project %s to no project", restoreNamespaceList, destProjectList[1]))
+			log.Infof("Verifying if restore is successful after moving destination namespaces to no project")
+			err = restoreSuccessCheck(restoreName, orgID, maxWaitPeriodForRestoreCompletionInMinute*time.Minute, restoreJobProgressRetryTime*time.Minute, ctx)
+			dash.VerifyFatal(err, nil, fmt.Sprintf("Verifying restore %s", restoreName))
```
Comment (mkoppal-px) [inline tests/backup/backup_rancher_test.go:434]: Mention the names of all the namespaces and restores for all these log statements

**Review by mkoppal-px — APPROVED** [review]
LGTM except for one last minor log related comment.

File: tests/backup/backup_rancher_test.go (hunk 6 above)
Comment (snigdha-px) [reply tests/backup/backup_rancher_test.go:434]: done

**Pull request merged**